### PR TITLE
fix: policy rules cannot be updated during upgrade

### PR DESCRIPTION
**Pull request by wansir**
Signed-off-by: hongming <talonwan@yunify.com>

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

fix policy rules cannot be updated during upgrade

### CODE REVIEW

**Comment by wansir** [comment]
cc @zryfish @runzexia

File: pkg/controller/namespace/namespace_controller.go [hunk 1]
```diff
@@ -228,6 +227,15 @@ func (r *ReconcileNamespace) checkAndCreateRoles(namespace *corev1.Namespace) er
 				return err
 			}
 		}
+		if !reflect.DeepEqual(found.Rules, role.Rules) {
+			found.Rules = role.Rules
+			err := r.Update(context.TODO(), found)
+			log.Info("Updating default role", "namespace", namespace.Name, "role", role.Name)
```
[the remark this answers is not in the document]
Comment (wansir) [reply pkg/controller/namespace/namespace_controller.go]: fixed

File: pkg/controller/namespace/namespace_controller.go [hunk 2]
```diff
@@ -208,6 +193,35 @@ func (r *ReconcileNamespace) Reconcile(request reconcile.Request) (reconcile.Res
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileNamespace) isControlledByWorkspace(namespace *corev1.Namespace) (bool, error) {
+
+	workspaceName := namespace.Labels[constants.WorkspaceLabelKey]
+
+	// without workspace label
+	if workspaceName == "" {
+		log.V(6).Info("deleting role bindings", "namespace", namespace.Name)
```
Comment (zryfish) [inline pkg/controller/namespace/namespace_controller.go]: use klog

File: pkg/controller/namespace/namespace_controller.go [hunk 3]
```diff
@@ -208,6 +193,35 @@ func (r *ReconcileNamespace) Reconcile(request reconcile.Request) (reconcile.Res
...
+		err := r.Delete(context.TODO(), adminBinding)
+		if err != nil && !errors.IsNotFound(err) {
+			log.Error(err, "deleting default role binding", "namespace", namespace.Name, "role binding", adminBinding.Name)
+			return false, err
+		}
+		viewerBinding := &rbac.RoleBinding{}
+		viewerBinding.Name = viewer.Name
+		viewerBinding.Namespace = namespace.Name
+		err = r.Delete(context.TODO(), viewerBinding)
+		if err != nil && !errors.IsNotFound(err) {
```
Comment (zryfish) [inline pkg/controller/namespace/namespace_controller.go]: don't we need to delete role operator?

File: pkg/controller/namespace/namespace_controller.go [hunk 4]
```diff
@@ -208,6 +193,35 @@ func (r *ReconcileNamespace) Reconcile(request reconcile.Request) (reconcile.Res
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileNamespace) isControlledByWorkspace(namespace *corev1.Namespace) (bool, error) {
```
Comment (zryfish) [inline pkg/controller/namespace/namespace_controller.go]: function name and what this function does doesn't match, `isControllerByWorkspace` should just test on a namespace to see if it belongs a workspace, but now it also does cleanup work.

File: pkg/controller/namespace/namespace_controller.go (hunk 3 above)
Comment (wansir) [reply pkg/controller/namespace/namespace_controller.go]: operator is not controlled by workspace

**Pull request merged**